### ISSUE TITLE
Fix S2583/S2589 FP: On null check if instance is used in try block

### ISSUE THREAD

**Issue by tsleegers** [SonarSource/sonar-dotnet#2590]
### Description

A S2583 is detected when checking for null when a variable is assinged using a function and the object is used inside a try block. Similarly a S2589 is detected when checking for not null when a variable is assinged using a function and the object is used inside a try block

### Repro steps

```
static class Program
{
    static void Main()
    {
        var foo = GetFoo(Guid.Empty);

        try
        {
            foo.Write();
        }
        catch
        {
            // Do nothing
        }

        if (foo == null) // S2583
        {
            Console.WriteLine("Foo is null");
        }

        if (foo != null) // 2589
        {
            Console.WriteLine("Foo is not null");
        }
    }

    static Foo GetFoo(Guid id)
    {
        return id != Guid.Empty ? new Foo(id) : null;
    }

    class Foo
    {
        private Guid Id { get; }

        public Foo(Guid id)
        {
            Id = id;
        }

        public void Write()
        {
            Console.WriteLine(Id);
        }
    }
}
```

### Expected behavior

No warnings

### Actual behavior

A S2583 warning on `if (foo == null)` and a 2589 warning on `if (foo != null)`

### Related information

* SonarC# Version 7.16.0.8981
* Visual Studio Version 16.1.5
* .NET Core 2.2

**Comment by tsleegers** [comment]
[SonarAlwaysTrueTest.zip](https://github.com/SonarSource/sonar-dotnet/files/3505815/SonarAlwaysTrueTest.zip)

**Comment by tsleegers** [comment]
Seems to be introduced in 7.16.0.8981. Earlier versions don't give a warning.

**Comment by andrei-epure-sonarsource** [comment]
thanks a lot for the detailed reproducer, @tsleegers - I confirm the FP

**Comment by andrei-epure-sonarsource** [comment]
Adding another repro [from community ](https://community.sonarsource.com/t/try-catch-confuses-detection-of-boolean-changes-s2589/13742)

```
bool blip = false;
try
{
	var rnd = new System.Random();
	if (rnd.Next() % 100 > 50) throw new Exception();
	blip = true;
}
catch { /* Ignore */  }

if (blip) Console.WriteLine("blip");
```

**Comment by jackerman88** [comment]
Similarly, if a variable is assigned to null outside of a try block, assigned a value inside the try block, and then checked for null in the catch block, S2583 is raised:

```
string test = null;

try 
{
    test = "something";
}
catch
{
    if (test != null) // Raises S2583
    {
        DoSomething();
    }
}

**Comment by s-kip** [comment]
Maybe also similarly:

    bool valid = false;
    Exception exception = null;

    foreach (int i in possibleNumbers)
    {
        try
        {
            ValidateNumber(i);
            valid = true;
        }
        catch (ArgumentOutOfRangeException ex)
        {
            exception = ex;
        }
    }
    if (!valid && exception != null) // Raises S2583 "always evaluate to 'false'" on "exception != null"
    {
        throw exception;
    }

With ValidNumber being:
    
    private static void ValidateNumber(int i)
    {
        if (i > 1) throw new ArgumentOutOfRangeException(nameof(i));
    }

So the first execution of `ValidateNumber` inside the foreach works fine and `valid` is set to `true`. The second execution throws the exception and `exception` is set in the `catch`-Block.

Removing the check "exception != null" raises S2259 "Null pointers should not be dereferenced" on "throw exception". That is only true if `possibleNumbers` contains no element.

BTW: none of these warnings are shown by the SonarLint-extension (v 4.12.0.0) for Visual Studio 2019 (16.3.6)
